MLS.Agent/NativeAssemblyLoadHelper.cs

```diff
@@ -1,50 +1,218 @@
+            // Where rid is: win, win-x64, win-x86, osx-x64, linux-x64 etc ...
+            IEnumerable<string> ProbingRids()
+            {
+                var processArchitecture = System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture;
+                var baseRid =
+                    System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "win" :
+                    System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? "osx" :
+                    "linux";
+
+                var platformRid =
+                    processArchitecture == Architecture.X64 ? "-x64" :
+                    processArchitecture == Architecture.X86 ? "-x86" :
+                    processArchitecture == Architecture.Arm64 ? "-arm64" :
+                    "arm";
+
+                yield return baseRid + platformRid;
+                yield return baseRid;
+                yield return "any";
+            }
         }
 
-        private AssemblyLoadEventHandler AssemblyLoaded(string assembly)
+        private void OnAssemblyLoad(object sender, AssemblyLoadEventArgs args)
         {
-            return (sender, args) =>
+            if (args.LoadedAssembly.IsDynamic ||
+                string.IsNullOrWhiteSpace(args.LoadedAssembly.Location))
+            {
+                return;
+            }
+
+            Logger.Log.Info("OnAssemblyLoad: {location}", args.LoadedAssembly.Location);
+
+            NativeLibrary.SetDllImportResolver(
+                args.LoadedAssembly,
+                (libraryName, assembly, searchPath) =>
+                {
+                    var ptr = IntPtr.Zero;
+                    if (_resolvers.TryGetValue(
+                        args.LoadedAssembly.Location,
+                        out var reference))
+                    {
+                        ptr = _probingPaths.SelectMany(di => ProbingPaths(di.FullName, libraryName).Select(dll => nativeLoader(dll))).FirstOrDefault();
+                    }
+                    if (ptr == IntPtr.Zero)
+                    {
+                        lock (globalProbingPaths)
+                        {
+                            ptr = globalProbingPaths.SelectMany(di => ProbingPaths(di.FullName, libraryName).Select(dll => nativeLoader(dll))).FirstOrDefault();
+                        }
+                    }
+                    return ptr;
+                });
+
+            IntPtr nativeLoader(string dll)
             {
-                if (args.LoadedAssembly.Location == assembly)
+                var ptr = IntPtr.Zero;
+                try
                 {
-                    NativeLibrary.SetDllImportResolver(args.LoadedAssembly, Resolve);
+                    ptr = NativeLibrary.Load(dll);
+                    Logger.Log.Info("NativeLibrary.Load({dll})", args.LoadedAssembly.Location);
+                    Console.WriteLine($"NativeLibrary.Load({dll})");
```

Comment:
Is this going to show up in the user's notebook output?

Reply:
Yes. The `Console.WriteLine` needs to be removed.